Taper implement

duncan-clark opened the pull request:
Allowed setting of taper params from the R side. Added tests

ifellows commented:
This may not be ready for a merge. The code looks fine, but the tapering implementation is more than just exposing the low level C model object. The main way the user will use tapering should probably be by calling `ernm`. What about the standard error calculation? What about default choices for tapering amount?

duncan-clark commented:
OK - I see, there is the likelihood function and model functions on the R side that are needed for tapered ERNM, so these need to be joined up. I'll take another crack at this.

The tests I added have an example of fitting a tapered model with the `ernm` function. I needed to pass the additional tapering args - for which there are currently no defaults, if someone wants to taper they need to know what they're doing and specify the parameters. We could add possible rules of thumbs if we wanted to.  Do you think this is a good way for the user to interact with tapered ERNMs?

ifellows commented:
Actually, looking at the tests I think the code is closer than I had originally thought. Here are my issues:

1. `ReModel` should be renamed. `Re` stands for reduced entropy, which was an early possible name for tapering. Perhaps `TaperedModel`?
2. ernm now has `modelArgs` and `modelParams` as parameters. This is confusing. I think just remove `modelParams` and `modelClass`. Would an interface that adds parameters `, tapering = FALSE, tapering_r = 3, beta = NULL` to `ernm()`make sense to you? Setting `beta` would override `tapering_r`.
3. I forget, are the betas here as defined in the paper with the inverse and square or is it the more straightforward just multiply times the term?
4. When implementing `tapering_r`, it uses the observed statistic to calculate the beta values. I suggest we do no tapering on a statistic's dimension if it is 0 or negative. Additionally, maybe consider adding 5 or 10 to the statistics so that we don't over-taper small networks or non-count statistics. The `r` calculation was designed for count statistics. Are there any stats where the suggested approach would over-taper? Thoughts @handcock ?
5. The `$info` functions in `ernm-models.R` need to reflect tapering. This will give the right standard errors.
6. How should `$info` be calculated for missing data models? Is the obvious approach of just adding the tapering adjustments to the full and missing components right? For that matter, does fitting work as expected with missing data?

duncan-clark commented:
Made some more changes to streamline:
On your comments:
1) Did lots of renaming
2) changed this to just modelArgs whcih still get passed to ernm.fit but now also contain the tapering param, modelType is retained and TaperedModel should be passed if tapering is needed.
3) I changed everything to the 'tau' parametrization i.e. just multiply by tau (tau = t for Tapering I'm guessing)
4) Done nothing on this - currently user must choose degree of tapering.
5) This should be reflected now - based on code in ergm.Tapered
6) Missing data not considered for ERNM for now (considering use case of ERNM for missing community prediction, this should probably change in the future)

ifellows commented:
Sorry for the delay, I've been on vacation.

This is looking very good. Here are my comments:

* There are still references to theta dependent in the code. I suggest removing these.
* Replace stray `=` assignments with `<-`.
* I still feel that the user experience is not quite there. It seems pretty non-intuitive how the user would specify a tapered model.  My understanding from the code is that for an r=3 they would have to do something like:
```
form<- ...
stats <- ernm::calculateStatistics(form)
    mod <- ernm(form, 
        modelArgs = list(modelClass = 'TaperedModel', centers=stats, tau = 1/(3^2*stats)),
        centers = stats)
```
This requires the user to a deeper understanding of the method and code base than is necessary (e.g. what is modelArgs? where are they passed to? What elements can the list have? What is a good value for centers? what is a good value for tau?). I suggest a user interface more like:
```
ernm <- function(formula,
                 tapered = TRUE,
                 tapering_r = 3,
                 modelArgs=list(),
                 likelihoodArgs = list(),
                 fullToggles=c("Compound_NodeTieDyad_Neighborhood","DefaultVertex"),
                 missingToggles=c("Compound_NodeTieDyadMissing_NeighborhoodMissing","VertexMissing"),
                 nodeSamplingPercentage=0.2,
                 modelType=NULL,
                 ...)
```
So that the way the user would specify a tapered model would be:
```
mod <- ernm(form)
```
Internally, the `ernm` function would set the appropriate `modelArgs` (recommendation: `centers=stats, tau= ifelse(stats<=0, Inf, 1 / (tapering_r ^2 * (stats + 5))`). A knowledgable user could override this by having specified `modelArgs` directly. 

I'll leave it to you whether `tapered` should be true by default. I have been leaning more and more toward tapering by default, especially in an ernm type situation where degeneracy is even more of an issue. Perhaps tapering can be the big driver for uptake of the package. tapering in tapered.ERGM has languished in a difficult to install state. Needing tapering may drive people to use this package for standard ergm like things.

duncan-clark commented:
Agree on the above. I will work on changing the interface a little. Especially if tapering is going to be default, I will remove the gmmObjective likelihood and just add a tapering option to the ERNM likelihood, after all tapering with tau = 0 means no tapering.

I don't know much about the missingErnmLikeihood, but I assume it could be tapered also, I will leave as is for now and we can come back to it later.

duncan-clark commented:
Starting a new branch and PR, something is preventing compilation and I cannot find it after many hours of searching

ifellows commented on R/ernm-models.R:
```diff
@@ -82,6 +84,12 @@ MissingErnmModel <- function(observedSampler, unobservedSampler, ...){
+	model <- sampler$getModel()
+	if(class(model)[1] == 'Rcpp_UndirectedModel' | class(model)[1] == 'Rcpp_DirectedModel'){
+	    tapered <- FALSE
+	}else{
```
Flip this condition. Check if the underlying Rcpp model is a tapering model and if so set tapering <- TRUE. I'm not sure if every class that isn't UndirectedModel or DirectedModel will forever and always be a tapered model. Also use `inherits` not `class(...)[1] == ...`.

ifellows commented on R/ernm.R:
```diff
@@ -206,56 +236,83 @@ calculateStatistics <- function(formula){
 			modelType <- FullErnmModel
-		else
+	    else
 			modelType <- MissingErnmModel
```
If tapered==TRUE and there is missing data, we should currently throw an error.

ifellows commented:
@duncan-clark Looks good. Merge in at your convenience.
